### ISSUE TITLE
GrokFast: Accelerated Grokking by Amplifying Slow Gradients

### ISSUE THREAD

**Issue by redknightlois** [kozistr/pytorch_optimizer#244]
GrokFast: Accelerated Grokking by Amplifying Slow Gradients

> One puzzling artifact in machine learning dubbed grokking is where delayed generalization is achieved tenfolds of iterations after near perfect overfitting to the training data. Focusing on the long delay itself on behalf of machine learning practitioners, our goal is to accelerate generalization of a model under grokking phenomenon. By regarding a series of gradients of a parameter over training iterations as a random signal over time, we can spectrally decompose the parameter trajectories under gradient descent into two components: the fast-varying, overfitting-yielding component and the slow-varying, generalization-inducing component. This analysis allows us to accelerate the grokking phenomenon more than ×50 with only a few lines of code that amplifies the slow-varying components of gradients. The experiments show that our algorithm applies to diverse tasks involving images, languages, and graphs, enabling practical availability of this peculiar artifact of sudden generalization

Twitter thread: https://x.com/_ironjr_/status/1798733867303772607
Paper: https://arxiv.org/abs/2405.20233
Code: 
- https://github.com/ironjr/grokfast (as an add-on)
- https://github.com/lucidrains/grokfast-pytorch (integrated in the optimizer)
